main.py

```diff
@@ -1,90 +1,84 @@
-from utils.generate_speakers_dataset import generate_speakers_dataset
+from core import (
+    Pipeline,
+    Pyannote,
+    WavLM,
+    TitaNet,
+    Ecapa
+)
+from core.metrics import (
+    compute_eer, 
+    compute_min_dcf,
+    compute_far_frr,
+)
+from utils import make_dataset
+
 
+def get_label(file1: str, file2: str) -> int:
+    """
+    Return 0 if different speakers, 1 if same speakers.
+    """
+    def _get_name(x):
+        return x.split("/")[-2]
 
-def far_frr_scores(y_true, y_pred):
-    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
-    far = fp / (fp + tn)
-    frr = fn / (fn + tp)
-    return far, frr
+    return int(_get_name(file1) == _get_name(file2))
 
 
 def evaluate_pipeline(
     pipeline, 
-    speakers_datasets, 
-    threshold: float = 0.5
+    data, 
 ) -> pd.DataFrame:
 
-    # Create table
-    scores_df = pd.DataFrame(columns=["pipeline", "speaker", "far", "frr", "accuracy", "threshold"])
-
-    for speaker, data in tqdm.tqdm(speakers_datasets.items(), total=len(speakers_datasets)):
-        # Make predictions
-        y_true = [label for _, label in data]
-        y_pred = []
-        for pair, _ in tqdm.tqdm(data, total=len(data)):
-            similarity = pipeline(*(Audio(a) for a in pair))
-            y_pred.append(int(similarity > threshold))
-        # Calculate metrics
-        far_score, frr_score = far_frr_scores(y_true, y_pred)
-        accuracy = accuracy_score(y_true, y_pred)
-        # Update table
-        row = {
-            "pipeline": pipeline.name,
-            "speaker": speaker, 
-            "far": far_score, 
-            "frr": frr_score, 
-            "accuracy": accuracy, 
-            "threshold": threshold,
-        }
-        scores_df = scores_df.append(row, ignore_index=True)
-    
-    return scores_df
+    scores = []
+    labels = []
 
+    for file1, file2 in tqdm.tqdm(data, total=len(data)):
+        similarity = pipeline(file1, file2)
+        label = get_label(file1, file2)
+        scores.append(similarity)
+        labels.append(label)
 
-def main():
-    # Get speaker to samples map
-    dataset_dir = "./dataset"
-    speakers = glob.glob(os.path.join(dataset_dir, "*"))
-    speakers = speakers[:2]
-    speaker_to_samples_dict = { s: glob.glob(os.path.join(s, "*.wav")) for s in speakers }
+    ee_rate, thresh, fa_rate, fr_rate = compute_eer(scores, labels)
+    min_dcf = compute_min_dcf(fr_rate, fa_rate)
+    fa_score, fr_score = compute_far_frr(scores, labels, thresh)
+
+    result = {
+        "pipeline": pipeline.name,
+        "fa_score": fa_score,
+        "fr_score": fr_score,
+        "ee_rate": ee_rate,
+        "dcf": min_dcf, 
+        "threshold": thresh,
+    }
+    return result
 
-    # Generate datasets for evaluation for each speaker
-    speakers_datasets = generate_speakers_dataset(speaker_to_samples_dict)
 
-    total_df = pd.DataFrame()
+def main():
+
+    dataset = make_dataset("./dataset")
```

Comment:
we need to be sure that we generate the same dataset for each experiment. I guess we can dump it into the file at the first run and load it for rest of experiments